minoc/src/main/java/org/opencadc/minoc/DeleteAction.java

```diff
@@ -0,0 +1,111 @@
+*  OpenCADC project.                    OpenCADC.
+*
+*  OpenCADC is free software:           OpenCADC est un logiciel libre ;
+*  you can redistribute it and/or       vous pouvez le redistribuer ou le
+*  modify it under the terms of         modifier suivant les termes de
+*  the GNU Affero General Public        la “GNU Affero General Public
+*  License as published by the          License” telle que publiée
+*  Free Software Foundation,            par la Free Software Foundation
+*  either version 3 of the              : soit la version 3 de cette
+*  License, or (at your option)         licence, soit (à votre gré)
+*  any later version.                   toute version ultérieure.
+*
+*  OpenCADC is distributed in the       OpenCADC est distribué
+*  hope that it will be useful,         dans l’espoir qu’il vous
+*  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+*  without even the implied             GARANTIE : sans même la garantie
+*  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+*  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+*  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+*  General Public License for           Générale Publique GNU Affero
+*  more details.                        pour plus de détails.
+*
+*  You should have received             Vous devriez avoir reçu une
+*  a copy of the GNU Affero             copie de la Licence Générale
+*  General Public License along         Publique GNU Affero avec
+*  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+*  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+*                                       <http://www.gnu.org/licenses/>.
+*
+************************************************************************
+*/
+
+package org.opencadc.minoc;
+
+import java.net.URI;
+
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.inventory.db.ArtifactDAO;
+import org.opencadc.minoc.ArtifactUtil.HttpMethod;
+
+/**
+ * Interface with storage and inventory to delete an artifact.
+ *
+ * @author majorb
+ */
+public class DeleteAction extends ArtifactAction {
+    
+    private static final Logger log = Logger.getLogger(DeleteAction.class);
+
+    /**
+     * Default, no-arg constructor.
+     */
+    public DeleteAction() {
+        super(HttpMethod.DELETE);
+    }
+
+    /**
+     * Delete the artifact.
+     * @param artifactURI The identifier for the artifact. 
+     */
+    @Override
+    public Artifact execute(URI artifactURI) throws Exception {
+        ArtifactDAO dao = getArtifactDAO();
+        Artifact artifact = getArtifact(artifactURI, dao);
+        
+        dao.delete(artifact.getID());
+        log.debug("deleting from storage...");
+        getStorageAdapter().delete(artifact.storageLocation);
+        log.debug("deletedfrom storage");
+        
```

Comment:
When you delete an artifact, you have to also create a DeletedArtifactEvent and put it with the DeletedEventDAO. You need to do both in a transaction. The second DAO is instantiated using the copy constructor so they share config, data source, and transaction manager. 

So there are really 3 actions here and only 2 can be made atomic with a database transaction. This is the crux of the distributed transaction we discussed and the decision about ordering: what if the second action fails? What error does the service report to the user?

1. If you delete from inventory and then fail, the next file-validate could recover the Artifact (if the StorageAdapter impl returns artifactURI in the iterator). You would report that the delete failed, but the artifact would be gone for awhile, that would get propagated, the artifact would be recovered, that would get propagated... so that's ultimately a failure to delete that looks like it worked fior awhile. ughh.

2. If you delete from storage first and then fail, you have an artifact in iventory but no file in storage: later GETs will fail. You would report that the delete failed to the caller, but the artifact would still be there and in future you may recover or complete the deletion. The next file-validate could possibly schedule a new download (it would set the storageLocation to null so that file-sync would pick it up) but only if another site has a copy already... or it would delete the artifact if there are no other copies and at that time the DeletedArtifactEvent would propagate; until then, global transfer negotiation would keep handing out URLs that don't work.

For now, if you delete from inventory and fire event first (in transaction), delete from storage while txn is open, and rollback if anything fails we end up with a very small chance of scenario 2.

Reply:
done